ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -349,28 +349,66 @@ protected AstNode visitJoin(final Join node, final Void context) {
         );
       }
 
-      final ColumnRef leftJoinField = getJoinFieldName(
-          comparisonExpression,
-          left.getAlias(),
-          left.getDataSource().getSchema()
-      );
+      final Set<ColumnRef> colsUsedInLeft = new ExpressionAnalyzer(analysis.getFromSourceSchemas())
+          .analyzeExpression(comparisonExpression.getLeft(), false);
+      final Set<ColumnRef> colsUsedInRight = new ExpressionAnalyzer(analysis.getFromSourceSchemas())
+          .analyzeExpression(comparisonExpression.getRight(), false);
 
-      final ColumnRef rightJoinField = getJoinFieldName(
-          comparisonExpression,
-          right.getAlias(),
-          right.getDataSource().getSchema()
-      );
+      final SourceName leftSourceName = getOnlySourceForJoin(
+          comparisonExpression.getLeft(), comparisonExpression, colsUsedInLeft);
+      final SourceName rightSourceName = getOnlySourceForJoin(
+          comparisonExpression.getRight(), comparisonExpression, colsUsedInRight);
+
+      final boolean flipped = leftSourceName.equals(right.getAlias());
+      if (!validJoin(flipped, left.getAlias(), right.getAlias(), leftSourceName, rightSourceName)) {
+        throw new KsqlException(
+            "Each side of the join must reference exactly one source and not the same source. "
+                + "Left side references " + leftSourceName
+                + " and right references " + rightSourceName
+        );
+      }
 
       analysis.setJoin(new JoinInfo(
-          leftJoinField,
-          rightJoinField,
+          flipped ? comparisonExpression.getRight() : comparisonExpression.getLeft(),
+          flipped ? comparisonExpression.getLeft() : comparisonExpression.getRight(),
           joinType,
           node.getWithinExpression()
       ));
 
       return null;
     }
 
+    private boolean validJoin(
```

Comment:
it does check because flipped is represents exactly this (flipped is true only if the left expression is the right data source) but I agree your logic is much cleaner :)